pkg/edit/editor_test.go

```diff
@@ -37,6 +37,37 @@ func TestEditor_DoesCommandWithLeadingSpaceToHistory(t *testing.T) {
 	testCommands(t, f.Store /* no commands */)
 }
 
+func TestEditor_FiltersCommandBeforeAddingToHistory(t *testing.T) {
+	noHistory := "_EMPTY_"
+
+	test := func(c, input, exp string) {
+		f := setup(rc(c))
+		defer f.Cleanup()
+
+		feedInput(f.TTYCtrl, input)
+		f.Wait()
+
+		if exp == noHistory {
+			testCommands(t, f.Store)
+		} else {
+			testCommands(t, f.Store, exp)
+		}
+	}
+
+	// item is not a callable, invalid
+	test("edit:add-cmd-filters = [$false]", "echo\n", "echo")
+	// function has no output, invalid
+	test("edit:add-cmd-filters = [[_]{}]", "echo\n", "echo")
+	test("edit:add-cmd-filters = [[_]{ put $true }]", "echo\n", "echo")
+	test("edit:add-cmd-filters = [[_]{ put $false }]", "echo\n", noHistory)
```

Comment:
IIUC these will write error messages to stderr when run. Do one the following:

* Make `callFilters` use a mutable global, instead of `os.Stderr`, and assign it to `eval.DevNull` during testing.

* Even better, assign the global to a pipe and check the output, and test the error message too. This probably creates a bit too much work for the sake of the PR so I'm OK with the lazier approach above.

Reply:
As the error output is an arbitrary string, it doesn't seem stable enough to match on it. Maybe we could have `diag.Complainf` use the global value internally.

How about a separated PR to improve the test for `callHooks` and `callFilters`?